Documentation: Update Admission Section (#5677)
kubernetes-sigs/kueue: site/content/en/docs/concepts/_index.md

Quota reservation is sometimes referred to as _workload scheduling_ or _job scheduling_,
but it should not to be confused with [pod scheduling](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/).

### Admission

_Admission_ is the process of allowing a Workload to start (Pods to be created). Kueue uses a two-step admission cycle for Workload scheduling: 

- Quota Reservation : When a Workload is submitted, it enters a LocalQueue first. This LocalQueue points to a ClusterQueue which is responsible for managing the available resources. The Kueue scheduler checks if the resources (CPU, memory, GPUs, etc.) requested by the Workload can be satisfied using the targeted ClusterQueue's available quota and resource flavors. If the quota is available, the resources are reserved for this workload and other Workloads are prevented from using the same resources. 

- Admission Checks: After the quota is reserved, Kueue executes all AdmissionChecks configured in the ClusterQueue concurrently. These are pluggable AdmissionChecks controllers that can perform certain validations such as policy checks, compliance, etc.

> These checks can be external or internal and determine if additional criteria are met before the Workload is admitted. The Workload is admitted once all its [AdmissionCheckStates](site/content/en/docs/concepts/admission_check.md) are mark `Ready`.

Author: @PBundyra should I remove Quota Reservation here as it's already mentioned in the Glossary section [here](https://kueue.sigs.k8s.io/docs/concepts/#quota-reservation) (above Admission) and directly start with the Admission Checks?

<h4> Example: Provisioning AdmissionCheck </h4>

Earlier, Kueue admissions were mainly based on quota checks- if sufficient quota existed, the workload is admitted. However, just because the quota (logical resource) is available, doesn't mean there are enough actual(physical) compute resources in the cluster to schedule all pods successfully. This is critical in cloud environments where autoscaling provides nodes on-demand. 

> [ProvisioningRequest AdmissionCheck](keps/1136-provisioning-request-support/README.md) handles this issue by ensuring that the physical resources can be provisioned before admitting Workloads.

Reviewer: Can we fix the link please

Reviewer 2: Yes, it is much better to provide documentation to other doc pages. KEPs we only reference as "supplementary materials", or "see also".

Lets understand this with a real-world usage - GPU Workload:

Scenario: *AI training job needing 16 GPUs*

- Step 1 (Quota Check): ClusterQueue has 32 GPU quota available, since requested is quota available, 16 GPUs are reserved.

- Step 2 (Capacity Check): A ProvisioningRequest is created for 16 GPUs. Cluster Autoscaler checks cloud provider GPU inventory and initiates scaling of 4x GPU nodes (4 GPUs each). It sets `Provisioned=true` when nodes ready.

Kueue sees the `Provisioned=true` proceeds to mark the AdmissionCheck `Ready` and admits workload

Outcome:
*Job starts immediately with all 16 GPUs available.*

<h4>Failure Handling: </h4>

> If the capacity check fails, incase of temporary issues like cloud capacity shortages, exponential backoff retries is triggered while maintaining the workload's queue position and reserved quota. For permanent failures, the AdmissionCheck is marked Rejected, quota is released, and the workload is evicted, requiring user resubmission.

Reviewer: Indeed holding reserved quota was what we did in the past, but we don't support it anymore. Currently if the admission check fails we release the quota and requeue the Workload

<!-- A Workload is admitted when it has a Quota Reservation and all its [AdmissionCheckStates](/docs/concepts/admission_check)

> are `Ready`. -->

Reviewer: Is this relevant? If no, can we remove it?

### Quota Reservation

_Quota reservation_ is the process during through which the kueue scheduler locks the resources needed by a workload within the targeted
[ClusterQueues ResourceGroups](/docs/concepts/cluster_queue/#resource-groups)
[ClusterQueues ResourceGroups](/docs/concepts/cluster_queue#resource-groups)

Quota reservation is sometimes referred to as _workload scheduling_ or _job scheduling_,
but it should not to be confused with [pod scheduling](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/).

### Admission

_Admission_ is the process of allowing a Workload to start (Pods to be created). A Workload
is admitted when it has a Quota Reservation and all its [AdmissionCheckStates](/docs/concepts/admission_check)

> are `Ready`.

Reviewer 3: The diagrams and code snippets might be useful. However, in this PR, let us focus on just preparing describing page. You could learn what to look for on the concept page on the Preemption page style: https://kueue.sigs.k8s.io/docs/concepts/preemption/

> are `Ready`.

Author: Got it👍

> - Admission Checks: After the quota is reserved, Kueue executes all [AdmissionChecks](/docs/concepts/admission_check) configured in the ClusterQueue concurrently. These are pluggable AdmissionChecks controllers that can perform validations such as policy checks, compliance, etc.

Reviewer: ```suggestion
- Admission Checks: After the quota is reserved, Kueue executes all [AdmissionChecks](/docs/concepts/admission_check) configured in the ClusterQueue concurrently. These are pluggable controllers that can perform validations such as policy checks, compliance, etc.
```

These checks can be external or internal and determine if additional criteria are met before the Workload is admitted. The Workload is admitted once all its [AdmissionCheckStates](/docs/concepts/admission_check/#admissioncheckstates) are marked `Ready`.

<h4> Example: Provisioning AdmissionCheck </h4>

> Without AdmissionChecks or [TopologyAwareScheduling](docs/concepts/topology_aware_scheduling/), Kueue admissions were mainly based on quota checks - if sufficient quota existed, the workload was admitted. While quota reservation ensures logical resource availability, it doesn't guarantee physical resources exist to schedule all pods successfully. The [ProvisioningRequest AdmissionCheck](/docs/admission-check-controllers/provisioning/) addresses this in cloud environments.

Reviewer: ```suggestion
Without AdmissionChecks or [TopologyAwareScheduling](docs/concepts/topology_aware_scheduling/), Kueue admissions were mainly based on quota checks - if sufficient quota existed, the Workload was admitted. While quota reservation ensures logical resource availability, it doesn't guarantee physical resources exist to schedule all Pods successfully. The [ProvisioningRequest AdmissionCheck](/docs/admission-check-controllers/provisioning/) addresses this in cloud environments.
```

- Capacity Guarantee: This step uses ProvisioningRequest and [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) to verify physical resource availability. 
  - The Kueue controller creates a ProvisioningRequest object by attaching the Workload's PodTemplates(optionally merged via [PodSetMergePolicy](/docs/admission-check-controllers/provisioning/#podset-merge-policy)) , applying [ProvisioningRequestConfig](/docs/admission-check-controllers/provisioning/#provisioningrequest-configuration) settings, and setting owner reference to Workload.
  - Cluster Autoscaler receives ProvisioningRequest, checks actual cluster capacity, triggers scaling if needed and updates ProvisioningRequest status with this possible states: 
    - `Provisioned=true`: Capacity guaranteed
    - `Provisioned=false`: Scaling in progress (Kueue surfaces ETA via Kubernetes events)
    - `Failed=true`: Cannot provision

>     - `BookingExpired=true`: Treated same as `Failed=true`  

Reviewer: I think it's useful to have the documentation about states of ProvisioningRequest, but let's make sure it doesn't confuse users. My proposal:

```
    - `Provisioned=false`: Provisioning in progress
    - `Provisioned=true`: CA provisioned the capacity and it's ready to use
    - `Failed=true`: CA couldn't provision the capacity
    - `BookingExpired=true`: CA stopped booking the capacity, it will scale down if there are no Pods running on it
    - `CapacityRevoked=true`: CA revokes the capacity, if a Workload is running on it, it will be evicted
```

    - `CapacityRevoked=true`: Evicts workloads with retry support 

Let's understand this with a real-world usage - GPU Workload:

Scenario: *AI training job requiring 16 GPUs*

- Step 1 (Quota Reservation): ClusterQueue has 32 GPU quota available. Kueue reserves 16 GPUs from this quota.

- Step 2 (Admission Check): Kueue creates a ProvisioningRequest requesting for 16 GPUs. 
  - Cluster Autoscaler checks cloud provider GPU inventory and initiates scaling of 4x GPU nodes (4 GPUs each). It sets `Provisioned=true` when nodes are ready.

  - Kueue sees the `Provisioned=true` proceeds to mark the AdmissionCheck `Ready` and admits workload

Outcome:
*Job starts immediately with all 16 GPUs available.*

<h4> Failure Handling: </h4> 

> - If the admission check fails due to temporary issues (e.g., cloud capacity shortages), the system releases the reserved quota, requeues the workload, and triggers exponential backoff retries.

Reviewer: Can we reference `retryStrategy` here? https://kueue.sigs.k8s.io/docs/admission-check-controllers/provisioning/#provisioningrequestconfig

Kueue creates new ProvisioningRequest with `-attempt<N>` suffix each retry.

> - For permanent failures the AdmissionCheck is marked `Rejected` and the workload is evicted, requiring user resubmission.
> 
> {{% alert title="Note" color="primary" %}} The `KeepQuotaForProvReqRetry` feature gate (v0.9) temporarily preserved quota during retries, but current implementations ALWAYS release quota and requeue workloads on failure.{{% /alert %}}

Reviewer: ```suggestion
- For permanent failures the AdmissionCheck is marked `Rejected`, the Workload is evicted and the quota it reserved is released. The Workload gets deactivated and to requeue it, a user needs to set `.status.active` field to `true`.
```

> {{% alert title="Note" color="primary" %}} The `KeepQuotaForProvReqRetry` feature gate (v0.9) temporarily preserved quota during retries, but current implementations ALWAYS release quota and requeue workloads on failure.{{% /alert %}}

Reviewer: Since this is deprecated I think we can omit documenting that

Author: Okay
